lib/checkunusedvar.cpp

```diff
@@ -1426,27 +1427,74 @@ void CheckUnusedVar::unusedStructMemberError(const Token *tok, const std::string
 bool CheckUnusedVar::isRecordTypeWithoutSideEffects(const Type* type)
 {
     // a type that has no side effects (no constructors and no members with constructors)
-    /** @todo false negative: check constructors for side effects */
-
     const std::pair<std::map<const Type *,bool>::iterator,bool> found=mIsRecordTypeWithoutSideEffectsMap.insert(
                 std::pair<const Type *,bool>(type,false)); //Initialize with side effects for possible recursions
     bool & withoutSideEffects=found.first->second;
     if (!found.second)
         return withoutSideEffects;
 
-    if (type && type->classScope && type->classScope->numConstructors == 0 &&
-        (type->classScope->varlist.empty() || type->needInitialization == Type::NeedInitialization::True)) {
-        for (std::vector<Type::BaseInfo>::const_iterator i = type->derivedFrom.begin(); i != type->derivedFrom.end(); ++i) {
-            if (!isRecordTypeWithoutSideEffects(i->type)) {
-                withoutSideEffects=false;
-                return withoutSideEffects;
+    // unknown types are assumed to have side effects
+    if (!type || !type->classScope)
+    {
+        withoutSideEffects = false;
+        return withoutSideEffects;
+    }    
+
+    // non-empty constructors -> possible side effects
+    if (type->classScope->numConstructors != 0)
+    {
+        for (std::list<Function>::const_iterator i = type->classScope->functionList.begin();
+            i != type->classScope->functionList.end(); ++i)
+        {
+            if (!i->isConstructor())
+            {
+                continue;
+            }
+            if (!i->isBodyEmpty())
+            {
+                withoutSideEffects = false;
+                return withoutSideEffects;    
+            }
+        }
+    }
+
+    // check each variable type for side effects
+    if (!(type->classScope->varlist.empty() || type->needInitialization == Type::NeedInitialization::True))
```

Comment:
this condition is obfuscated imho. This is logically the same and does not have an outer not:

    if (!type->classScope->varlist.empty() && type->needInitialization != Type::NeedInitialization::True)

The first condition is redundant because the conditional loop will do nothing if the varlist is empty.